Added swagger for Azure container registry

djyou opened the pull request:
<!-- Reviewable:start -->

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/azure/azure-rest-api-specs/549)

<!-- Reviewable:end -->

djyou committed:
Added swagger for Azure container registry

azurecla commented:
Hi __@djyou__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <p>
        It looks like you're working at Microsoft (doyou). If you're full-time, we DON'T require a contribution license agreement.
    </p>
    <p>
        If you are a vendor, DO please sign the electronic contribution license agreement. It will take 2 minutes and there's no faxing! https://cla.azure.com.
    </p>

TTYL, AZPRBOT;

amarzavery commented:
@djyou 

- Please follow the instructions and snapshots over [here](https://github.com/Azure/openapi-lint-extension#installation-instructions) to install the linter and make sure you resolve all the warnings and errors.
- Please read the good swagger patterns that we recommend from [here](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-good-patterns.md) and make sure that they are followed when you author the swagger

Please make sure that the above two steps are done before we start the review.

djyou commented:
/cc @SajayAntony @sivagms

amarzavery commented on arm-containerregistry/2016-06-27-preview/swagger/containerregistry.json:
```diff
@@ -0,0 +1,361 @@
+          "200": {
+            "description": "OK",
+            "schema": {
+              "$ref": "#/definitions/Object"
```
Why would you return an "Object"? Please stop using swashbuckle and hand author your swagger. Swashbuckle generates bad swaggers that are difficult to consume.

amarzavery commented on arm-containerregistry/2016-06-27-preview/swagger/containerregistry.json:
```diff
@@ -0,0 +1,361 @@
+        "tags": [
+          "Registries"
+        ],
+        "operationId": "Registries_Delete",
```
missing description

amarzavery commented on arm-containerregistry/2016-06-27-preview/swagger/containerregistry.json:
```diff
@@ -0,0 +1,361 @@
+          "200": {
+            "description": "OK",
+            "schema": {
+              "$ref": "#/definitions/RegistryParameters"
```
the name of the definition should be "Registry" and not "RegistryParameters"

amarzavery commented on arm-containerregistry/2016-06-27-preview/swagger/containerregistry.json:
```diff
@@ -0,0 +1,361 @@
+        "tags": [
+          "Registries"
+        ],
+        "operationId": "Registries_Create",
```
missing description

amarzavery commented on arm-containerregistry/2016-06-27-preview/swagger/containerregistry.json:
```diff
@@ -0,0 +1,361 @@
+            "$ref": "#/parameters/SubscriptionIdParameter"
+          },
+          {
+            "name": "resourceGroup",
```
general comment: missing description

amarzavery commented on arm-containerregistry/2016-06-27-preview/swagger/containerregistry.json:
```diff
@@ -0,0 +1,361 @@
+        "tags": [
+          "Registries"
+        ],
+        "operationId": "Registries_Update",
```
Registries_Patch. All the Patch operations should have the verb of the operation as Patch.

amarzavery commented on arm-containerregistry/2016-06-27-preview/swagger/containerregistry.json:
```diff
@@ -0,0 +1,361 @@
+          "200": {
+            "description": "OK",
+            "schema": {
+              "$ref": "#/definitions/Object"
```
Looks incorrect.

amarzavery commented on arm-containerregistry/2016-06-27-preview/swagger/containerregistry.json:
```diff
@@ -0,0 +1,361 @@
+          "200": {
+            "description": "OK",
+            "schema": {
+              "$ref": "#/definitions/ResourceList[RegistryParameters]"
```
RegistryParametersResourceList. Please make sure that square brackets are not present in the generated model definitions

amarzavery commented on arm-containerregistry/2016-06-27-preview/swagger/containerregistry.json:
```diff
@@ -0,0 +1,361 @@
+          "200": {
+            "description": "OK",
+            "schema": {
+              "$ref": "#/definitions/ResourceList[RegistryParameters]"
```
same as above.

amarzavery commented on arm-containerregistry/2016-06-27-preview/swagger/containerregistry.json:
```diff
@@ -0,0 +1,361 @@
+    "RegistryParameters": {
+      "type": "object",
+      "properties": {
+        "id": {
```
All the Resource specific properties must be defined in the "Resource" class. Registry parameters must inherit `"allOf"` from the "Resource" model. Please take a look at the Compute or Keyvault or any other spec as an example. 

Is location a required property of the Resource in your case?

amarzavery commented on arm-containerregistry/2016-06-27-preview/swagger/containerregistry.json:
```diff
@@ -0,0 +1,361 @@
+          "readOnly": true
+        },
+        "properties": {
+          "$ref": "#/definitions/RegistryProperties"
```
`"x-ms-client-flatten": true`

amarzavery commented on arm-containerregistry/2016-06-27-preview/swagger/containerregistry.json:
```diff
@@ -0,0 +1,361 @@
+        },
+        "creationDate": {
+          "format": "date-time",
+          "type": "string"
```
the description should state that the value will be in ISO8601 format.

amarzavery commented on arm-containerregistry/2016-06-27-preview/swagger/containerregistry.json:
```diff
@@ -0,0 +1,361 @@
+    "RegistryProperties": {
+      "type": "object",
+      "properties": {
+        "storageAccount": {
```
what are the set of 
- "required"
- "readOnly"

properties in the model definition? Please describe them accurately.

amarzavery commented on arm-containerregistry/2016-06-27-preview/swagger/containerregistry.json:
```diff
@@ -0,0 +1,361 @@
+        }
+      }
+    },
+    "Object": {
```
Why do you need this model?

amarzavery commented on arm-containerregistry/2016-06-27-preview/swagger/containerregistry.json:
```diff
@@ -0,0 +1,361 @@
+    "ResourceList[RegistryParameters]": {
+      "type": "object",
+      "properties": {
+        "value": {
```
value should be a required property.

amarzavery commented on arm-containerregistry/2016-06-27-preview/swagger/containerregistry.json:
```diff
@@ -0,0 +1,361 @@
+    "SubscriptionIdParameter": {
+      "name": "subscriptionId",
+      "in": "path",
+      "description": "Gets subscription credentials which uniquely identify Microsoft Azure subscription.The subscription ID forms part of the URI for every service call.",
```
Keep the description short and simple
> "The Azure subscription id."

amarzavery commented on arm-containerregistry/2016-06-27-preview/swagger/containerregistry.json:
```diff
@@ -0,0 +1,361 @@
+        "tags": [
+          "Registries"
+        ],
+        "operationId": "Registries_List",
```
please add the extension "x-ms-pageable". You can find more information about this doscument over [here](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-extensions.md#x-ms-pageable). Everything is documented, so please make sure to read the documentation correctly and then model the extension as per your use case.

amarzavery commented on arm-containerregistry/2016-06-27-preview/swagger/containerregistry.json:
```diff
@@ -0,0 +1,361 @@
+            "in": "body",
+            "required": true,
+            "schema": {
+              "$ref": "#/definitions/RegistryParameters"
```
make sure that the body parameter of the patch operation has all optional properties, no readonly properties and no defaults.

amarzavery reviewed (changes requested):
make sure to address the feedback before the spec can be merged. 
Please run the linter and also read the recommended patterns.

amarzavery commented on arm-containerregistry/2016-06-27-preview/swagger/containerregistry.json:
```diff
@@ -0,0 +1,361 @@
+{
```
General Comment: The operations, parameters, models, modelProperties should all be well documented by providing accurate description.

amarzavery commented:
@djyou - What is the status of this PR ? When can the review comments be addressed?

djyou commented:
Our RP is making several major changes and we are working on the swagger file. We will try to finish this ASAP.

amarzavery commented:
@djyou - Any further update?

djyou commented:
@amarzavery I believe we have addressed requested changes as much as we can. We appreciate any additional feedback. Thanks! /cc @SajayAntony

djyou commented:
/cc @JasonRShaver

amarzavery commented on arm-containerregistry/2016-06-27-preview/swagger/containerregistry.json:
```diff
@@ -0,0 +1,642 @@
+          "type": "string"
+        },
+        "adminUserEnabled": {
+          "description": "The boolean value that indicates whether admin user is enabled.",
```
Since this is optional. Please specify the default behavior in the description.
"The boolean value that indicates whether admin user is enabled. Default value is false."

djyou commented on arm-containerregistry/2016-06-27-preview/swagger/containerregistry.json:
```diff
@@ -0,0 +1,642 @@
+          "Operation"
+        ],
+        "description": "Checks whether the container registry name is available.",
+        "operationId": "Operation_CheckNameAvailability",
```
Our RP explicitly put this under Operation since the path goes to Microsoft.ContainerRegistry/checkNameAvailability rather than, for example, Microsoft.ContainerRegistry/registries/{registryName}. I have seen different patterns. For example, storage account put checkNameAvailability under StorageAccount, while CDN put it under a special NameAvailability: https://github.com/Azure/azure-rest-api-specs/blob/master/arm-cdn/2016-10-02/swagger/cdn.json#L1594 Is this a required change or can we keep it as operation?

anudeepsharma replied:
I would say change it as per Amar's recommendation, all azure apis should behave more or less in same way. With next CDN swagger changes we will try to change it to similar way.

---
In reply to: [83991770](https://github.com/Azure/azure-rest-api-specs/pull/549#discussion_r83991770) [](ancestors = 83991770)

djyou commented on arm-containerregistry/2016-06-27-preview/swagger/containerregistry.json:
```diff
@@ -0,0 +1,642 @@
+          "description": "Storage account endpoint URL.",
+          "type": "string"
+        },
+        "location": {
```
One of our requirements is that the registry has to be in the same location as the storage account. So either the request will have location or the RP will query storage account location using storage account name (also the reason why we want to have storage account name as a property for registry). We probably want to remove it in the future, since there is no guarantee the provided location is the actual location of the storage account, so we made it optional here.

djyou commented on arm-containerregistry/2016-06-27-preview/swagger/containerregistry.json:
```diff
@@ -0,0 +1,642 @@
+    "StorageAccountProperties": {
+      "description": "The storage account properties.",
+      "required": [
+        "name",
```
Right, accessKey and endpointUrl are enough to identify and access one storage account. Our RP needs to query storage account properties from name or resource id. We also want to support changing storage account for a registry, and in the future, map multiple storage accounts to one registry. It is just easier to manage with storage account name as required property.

djyou commented on arm-containerregistry/2016-06-27-preview/swagger/containerregistry.json:
```diff
@@ -0,0 +1,642 @@
+        "location": {
+          "description": "Resource location.",
+          "type": "string",
+          "readOnly": true
```
I read the documentation suggesting location to be required in Resource before making the changes and I was actually puzzled. When we create/update a registry, we use RegistryCreateParameters/RegistryUpdateParameters. The Registry definition (which inherits Resource) will only be used in response. In that sense, why do we want to make it required? We were following the same pattern from storage account and they didn't have location as a required property (https://github.com/Azure/azure-rest-api-specs/blob/master/arm-storage/2016-01-01/swagger/storage.json#L948). Futher, shouldn't location be treated in the same way as, for example, resource name, since once they are set, they cannot be changed, unless the resource is recreated? Shouldn't location be readOnly then?

amarzavery replied:
To give you an idea, this is the model I was envisioning for StorageManagement spec and I believe Container Registry will have a very similar model. 

- **CreateOrUpdate {PUT}** (body parameter StorageAccount) -> StorageAccount
- **Update {PATCH}** (body parameter MutableStorageAccount) -> StorageAccount
- **Get {GET}** () -> StorageAccount

![image](https://cloud.githubusercontent.com/assets/4521223/19531202/a80b74c8-95eb-11e6-9373-65be43abbb55.png)

![image](https://cloud.githubusercontent.com/assets/4521223/19530891/93b3fab4-95ea-11e6-9eb3-4d8498947989.png)


Here, visualize 
- `StorageAccount` as `ContainerRegistry`
    - `StorageAccountProperties` as `ContainerRegistryProperties`
- `MutableStorageAccount` as `ContainerRegistryUpdateParameters`
     - `MutableStorageAccountProperties` as `ContainerRegistryPropertiesUpdateParameters`


- **CreateOrUpdate {PUT}** (body parameter ConatinerRegistry) -> ConatinerRegistry
- **Update {PATCH}** (body parameter ConatinerRegistryUpdateParameters) -> ConatinerRegistry
- **Get {GET}** () -> ConatinerRegistry

What do you think about it?

djyou replied:
@amarzavery Thanks a lot for the detailed suggestion, this is great and we will definitely adopt this model and make improvements. I just want to make sure my understanding of the scenario is correct using the same model as input/output for PUT.

GET | PUT (or even PUT | PUT), if the user modified location (since location is required and not readOnly), the RP will throw an error. We don't seem to have a way to inform the user that location cannot be changed for an existing resource, or is this considered a common sense to Azure resources?

The pipeline scenario you mentioned earlier GET | UPDATE, the user can only do this by calling a PUT after GET, if the above scenario is enabled. We can't have GET | PATCH, since PATCH uses a different model as input and all properties are optional.

djyou commented on arm-containerregistry/2016-06-27-preview/swagger/containerregistry.json:
```diff
@@ -0,0 +1,642 @@
+        }
+      }
+    },
+    "StorageAccountBase": {
```
We used StorageAccount for a different purpose in our RP and here StorageAccountBase, as a response to the user, will only contain storage account name.

djyou commented on arm-containerregistry/2016-06-27-preview/swagger/containerregistry.json:
```diff
@@ -0,0 +1,642 @@
+          "Registries"
+        ],
+        "description": "Creates a new container registry with the specified parameters.",
+        "operationId": "Registries_Create",
```
Right, our RP will try to do an update if the resource already exists. The issue of making it CreateOrUpdate is that, create and update use different parameters (although technically update uses a subset of parameters from create). In our case, for example, location is a required property in create, but is forbidden to change in update. If the user happens to PUT the correct set of parameters for an existing resource, the RP will return success; otherwise, the RP still returns failure. What is the recommended pattern here?

amarzavery commented on arm-containerregistry/2016-06-27-preview/swagger/containerregistry.json:
```diff
@@ -0,0 +1,642 @@
+        "storageAccount"
+      ],
+      "properties": {
+        "tenantId": {
```
@djyou - `tenantId` is not sensitive information like storage account access keys; then why is `tenantId` not returned in the response, like adminUserEnabled ?
The reasoning behind asking this question is - 
We want to make it easy for the customer to update the registry. If we have the same model than it becomes easy to enable the pipeline scenario: `GET | UPDATE`; where the customer can do a GET, modify some properties of the object and then do a PUT. If tenantId is also returned in the response then it is very simple to operate.

djyou replied:
This is a reasonable use scenario, @mnltejaswini, can also we return tenantId? Actually, the response from GET is Registry, the parameters for PATCH and PUT are RegistryUpdateParameters and RegistryCreateParameters. We won't return storage account access key and endpoint url, but we also allow the users to update them. Besides, I doubt if this kind of pipeline will ever work. We were following the same pattern from StorageAccount, for example, [here ](https://github.com/Azure/azure-rest-api-specs/blob/master/arm-storage/2016-01-01/swagger/storage.json#L746)"sku" is readOnly, but [here ](https://github.com/Azure/azure-rest-api-specs/blob/master/arm-storage/2016-01-01/swagger/storage.json#L864)"sku" can also be updated. We also set all properties in Registry to be readOnly, that means the user cannot change them and send the updated registry back. The only way to do this is to use RegistryUpdateParameters. This also corresponds to the discussion on CreateOrUpdate, I think.

amarzavery replied:
We are planning to revamp the storage management spec. It has not been modeled accurately. So having that as a reference check is not ideal.

Yes, I was thinking about StorageAccount. As per the current model accessKey, endPointUrl and location are _write-only_ (that is they can be created and updated) but cannot be read. _Currently swagger does not provide a mechanism to specify write-only_. 
Actually, only **accessKey** contains sensitive information. What is the harm in sending the **endPointUrl** and **location** back in the response? 
Ideally StorageAccountBase  should have `{ name, location, endPointUrl }` as **required** properties.
StorageAccountProperties { accessKey (required) } should inherit from StorageAccountBase.

**RegistryProperties.storageAccount** should be of  type **StorageAccountProperties**. So when the service sends the response it will not have accessKeys, which I think should be fine as we do not validate on deserialization.

So having Registry as input to PUT and output of PUT, you will make customers life easy and open up the CreateOrUpdate() scenario on PUT. 

The input model for PATCH as an update is good.

djyou commented:
@amarzavery Thank you for your great feedback in the email and we have made suggested changes. However, as @sivagms suggested in the email, we will use the same Registry model as the input and output for PUT, so that our Create will actually be CreateOrUpdate, but we will return null for storage account accessKey in the response. Since VM also has the similar model for admin password, we consider it as acceptable. Besides, once we move to first party, we will remove accessKey and endpointUrl from storage account properties.

SajayAntony commented:
@JasonRShaver We want to get this spec reviewed and ironed out so that we can iterate on the RP this week.

amarzavery commented on arm-containerregistry/2016-06-27-preview/swagger/containerregistry.json:
```diff
@@ -0,0 +1,541 @@
+        "tags": [
+          "Registries"
+        ],
+        "description": "Creates a new container registry with the specified parameters.",
```
Creates or updates a new container registry with the specified parameters. Please make sure that the storage account that is provided in this operation is created in the same location in which the Registry will be created.

amarzavery commented:
LGTM

djyou committed:
Added regenerate credentials operation

1. Added regenerate credentials operation.
2. Updated storage account properties.

amarzavery commented on arm-containerregistry/2016-06-27-preview/swagger/containerregistry.json:
```diff
@@ -323,6 +323,46 @@
         "deprecated": false
       }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroup}/providers/Microsoft.ContainerRegistry/registries/{registryName}/regenerateCredentials": {
```
instead of resourceGroup the name of the path parameter should be resourceGroupName. This will be consistent with other specs.

djyou committed:
Renamed resourceGroup to resourceGroupName

SajayAntony commented:
@amarzavery We need to freeze on these changes this week. Do you think that would be reasonable. 
@djyou - If there are any more change we will have to handle them appropriately or maybe for GA but I would recommend this iteration gets merged once the RP is updated and we have the SDKs

hovsepm commented on arm-containerregistry/2016-06-27-preview/swagger/containerregistry.json:
```diff
@@ -0,0 +1,576 @@
+            }
+          }
+        },
+        "deprecated": false
```
>"deprecated": false [](start = 8, length = 19)

Why do you have this everywhere? False is default value anyway.

hovsepm commented on arm-containerregistry/2016-06-27-preview/swagger/containerregistry.json:
```diff
@@ -0,0 +1,576 @@
+        "deprecated": false,
+        "x-ms-pageable": {
+          "nextLinkName": null
+        }
```
Is there any specific reason to disable Pagination? Keep in mind that if at some point in the future server will return more than 1 page of results you will have to make a breaking change in all the SDK's.

hovsepm replied:
nextLink is not a required field and can be null in the server response. So no need to do any work on the server side. Specifying a filed name for nextLinkName will require you to modify the return value (adding a filed with that name) and will change method return type to IPage<YourType>.

amarzavery replied:
@hovsepm - That would be inconsistent with our primary goal "Swagger spec should accurately describe what is there on the wire". Going from non-pageable to pageable is a breaking change. Wouldn't it be correct to consume a breaking change in the API because of that (whenever that happens) ?

hovsepm replied:
@amarzavery Supporting pages should not be a breaking change on the SDK side IMO, it is a server side enhancement. The next link field is optional always, so the model will still describe accurately what is on the wire :).

hovsepm commented on arm-containerregistry/2016-06-27-preview/swagger/containerregistry.json:
```diff
@@ -0,0 +1,576 @@
+        },
+        "deprecated": false,
+        "x-ms-pageable": {
+          "nextLinkName": null
```
>"nextLinkName": null [](start = 10, length = 20)

Same comment. Are you going to lock number of container registries at "no more than one serve page" in the subscription?

SajayAntony replied:
@hovsepm can the swagger have  `"nextLinkName": "nextLink"`   does the RP need to return the field since it doesn't support pagenation at this point.

amarzavery replied:
The RP must send the "nextLink" property in the response only when there is more information to be provided in the nextPage. Otherwise the structure would be `{"value": [ . . . ] }`.

djyou committed:
Removed deprecated, updated nextLink

hovsepm commented on arm-containerregistry/2016-06-27-preview/swagger/containerregistry.json:
```diff
@@ -249,9 +244,8 @@
-        "deprecated": false,
         "x-ms-pageable": {
-          "nextLinkName": null
+          "nextLinkName": "nextLink"
```
Just adding this thing will not be enough. You need to modify RegistryListResult type as well. It should look like this:
```
"RegistryListResult": {
      "description": "The result of listing container registries.",
      "properties": {
        "value": {
          "type": "array",
          "items": {
            "$ref": "#/definitions/Registry"
          }
        },
        "nextLink": {
          "type": "string",
          "description": "Gets or sets the URL to get the next set of results."
        }
      }
    }
```

hovsepm commented:
@djyou please open you swagger in VS code with Swagger linter extension installed and fix all "missing description" warnings.

djyou commented:
@hovsepm We had used swagger linter to fix issues, is there anything we didn't catch in the current swagger?

hovsepm commented:
@djyou No, just noticed that RegistryListResult value field does not have any description and thought that you may forgot to run linter. Then please ignore my previous comment.

djyou committed:
Added nextlink to list results

hovsepm commented:
LGTM

amarzavery commented:
:shipit:

amarzavery commented on arm-containerregistry/2016-06-27-preview/swagger/containerregistry.json:
```diff
@@ -0,0 +1,572 @@
+        }
+      ],
+      "properties": {
+        "properties": {
```
one more thing. 
"properties" property of "Registry" must also be required. "RegistryProperties" is flattened. It has a required property "storageAccount". This must be the required property of "Registry" after flattening. Hence, it is very important to make "properties" of "Registry" required.

djyou replied:
RegistryProperties is flattened and it will not show up in the generated SDK. How does this matter? For example, the Python SDK generated will have
```
def __init__(self, location, storage_account, tags=None, admin_user_enabled=False):
```
which is correct as location and storage_account are the required properties.

amarzavery replied:
Ahh I see.. Since there is only one property it is working fine. I think if more than one property is required in the object that is supposed to be flattened then it will be a problem. Hence it is safer to mark properties required as well.

djyou replied:
I have an impression that `required` is redundant to `x-ms-client-flatten` because the object will go away anyway.

SajayAntony replied:
:shipit: 
@amarzavery can you or your team help merge this PR . We can open issues and track anything needed on top of this. Please assign to me or and @djyou for anything you need fixed.

amarzavery commented:
@vishrutshah -Do not publish a ruby gem for container registry right now after the rest-api-specs PR is merged.

AutorestCI commented:
No modification for NodeJS

AutorestCI commented:
No modification for Ruby

AutorestCI commented:
No modification for Python

vishrutshah commented:
@amarzavery Thanks for the heads-up. :) Just curious what are the reasons?

amarzavery commented:
@vishrutshah - The service is not fully ready yet. It will be tested in the new CLI and then we will get a green signal.
